stdlib/json.gr

```diff
@@ -0,0 +1,2157 @@
+/**
+ * JSON (JavaScript Object Notation) parsing and printing.
+ */
+
+module Json
+
+include "string"
+include "char"
+include "list"
+include "option"
+include "number"
+include "uint8"
+include "runtime/unsafe/wasmi32"
+include "runtime/unsafe/wasmi64"
+include "runtime/unsafe/wasmf32"
+include "runtime/unsafe/wasmf64"
+include "runtime/numberUtils"
+include "runtime/unsafe/tags"
+include "runtime/numbers"
+include "runtime/unsafe/memory"
+include "buffer"
+include "runtime/string" as RunTimeString
+include "runtime/numbers"
+include "runtime/unsafe/memory"
+include "runtime/dataStructures"
+include "runtime/bigint" as BI
+from RunTimeString use { toString as runtimeToString, getCodePoint }
+from Numbers use { coerceNumberToWasmI32 }
+from DataStructures use { tagSimpleNumber, newFloat64, untagSimpleNumber }
+
+// Primitive Offsets
+// TODO(703): Get these offsets from the runtime
+@unsafe
+let _INT64_BOXED_VALUE_OFFSET = 8n
+@unsafe
+let _Float64_BOXED_VALUE_OFFSET = 8n
+/**
+ * Data structure representing the result of parsing and the input of printing
+ * JSON.
+ *
+ * For example this object
+ * ```grain
+ * JSONObject([
+ *   ("currency", JSONString("€")),
+ *   ("price", JSONNumber(99.99)),
+ * ])
+ * ```
+ * is equivalent to the following JSON:
+ * ```json
+ * {"currency":"€","price":99.99}
+ * ```
+ *
+ * This data structure is semantically equivalent to the JSON format allowing
+ * mostly lossless round trips of printing and parsing. Exceptions to this are
+ * white spaces, multiple ways JSON allows escaping characters in strings and
+ * some edge-cases related to Grain's `Number` type.
+ *
+ * For example parsing the following JSON text will also result in the same
+ * `JSON` object as above.
+ * ```json
+ * {
+ *   "currency": "\u20ac",
+ *   "price": 99.99
+ * }
+ * ```
+ */
+provide enum JSON {
+  JSONNull,
+  JSONBoolean(Bool),
+  JSONNumber(Number),
+  JSONString(String),
+  JSONArray(List<JSON>),
+  // Note that JSONObject here is deliberately defined as a simple list of key value pair tuples as opposed
+  // to for example a Map in order to accomodate the fact that the ECMA-404 standard doesn't prohibit
+  // duplicate names in Objects. Such JSON should be representable by the JSON data structure for lossless
+  // processing. This also simplifies implementation by not requiring a purpose built data structure and
+  // has the benefit of List's immutability. It's a concious decision that sacrifices ease of use of the
+  // API for lossless handing of these edge cases with intention of later building more ergonomic APIs on a
+  // higher level of abstraction.
+  JSONObject(List<(String, JSON)>),
+}
+
+/**
+ * Represents errors for cases when a `JSON` object cannot be represente in the
+ * JSON format along with a human readable text message. This can happen when
+ * it contains number values `NaN`, `Infinity` or `-Infinity`.
+ */
+provide enum JSONToStringError {
```

Comment:
One of the things we're a bit inconsistent on is the casing of `JSON` vs `Json`. For acronyms in camel/pascal case, using the latter form is more clear, and I'd like Grain to take this as the official idiomatic format. @spotandjake Can you replace all identifiers that include `JSON` with `Json`? This does not include descriptions, just the identifiers.